### DISCUSSION TITLE
Setting a permanent password for MSSQL SA account

### DISCUSSION THREAD

**Discussion by Dazfl** [microsoft/aspire#4377] [Q&A] [answered]
I'm trying to add a MS SQL db to my Aspire project, but I would like to use a permanent password so I can check my db in MSSQL Management Studio (or Azure Data Studio).  I've tried to define it like below, but I keep getting a `Login failed for user 'sa'. Reason: Password did not match that for the login provided` error when my API tries to create the database.  Is this the correct way to define a password for the 'sa' account?

AppHost - Program.cs
```
var db = builder.AddSqlServer("test-sql-server", port: 54478)
    .WithEnvironment("MSSQL_SA_PASSWORD", "Test123!")
    .AddDatabase("test-db", "test");
 
var api = builder.AddProject<Projects.Test_Api>("test-api")
    .WithReference(db);
```
Api - Program.cs
```
builder.AddSqlServerDbContext<TestDbContext>("test-db");
```

**Comment by davidfowl** [comment, marked as answer]
This is where you use parameters in the application model. If you look at the overloads of [AddSqlServer](https://learn.microsoft.com/en-us/dotnet/api/aspire.hosting.sqlserverbuilderextensions.addsqlserver?view=dotnet-aspire-8.0), you'll see there's a password parameter that takes an `IResourceBuilder<ParameterResource>`.


```C#
var password = builder.AddParameter("sqlPass", secret: true);

var db = builder.AddSqlServer("test-sql-server", port: 54478, password: password)
    .AddDatabase("test-db", "test");
 
var api = builder.AddProject<Projects.Test_Api>("test-api")
    .WithReference(db);
```

The value for this parameter should be set in [user secrets](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-8.0&tabs=windows). Parameters are read from configuration in the format "Parameters:parametername", so the above would be "Parameters:sqlPass". 

You can learn more about parameters here https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/external-parameters
